flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedHaServicesWithLeadershipControl.java

```diff
@@ -36,7 +36,15 @@ public EmbeddedHaServicesWithLeadershipControl(Executor executor) {
         super(executor);
         this.testingCheckpointRecoveryFactory =
                 new PerJobCheckpointRecoveryFactory(
-                        n -> new EmbeddedCompletedCheckpointStore(),
+                        (maxCheckpoints, previous) -> {
+                            if (previous != null) {
+                                final EmbeddedCompletedCheckpointStore cast =
+                                        (EmbeddedCompletedCheckpointStore) previous;
```

Comment:
I am wondering whether we can make `PerJobCheckpointReocveryFactory` type safe by introducing a type parameter.

Reply:
👍 fixed